File: c/factorial.c
/*
 * Algorithm: Factorial (Recursive)
 * Description: Computes n! using recursion to demonstrate stack-based algorithms.
 * Usage: Compile with `gcc factorial.c -o factorial` and run `./factorial`.
 */
#include <stdio.h>

unsigned long long factorial(unsigned int n) {
    if (n == 0 || n == 1) {
        return 1ULL;
    }
    return n * factorial(n - 1);
}

int main(void) {
    unsigned int value = 5;
    unsigned long long result = factorial(value);
    printf("%u! = %llu\n", value, result);
    return 0;
}

File: rust/factorial.rs
// Algorithm: Factorial (Recursive)
// Description: Calculates n! recursively to illustrate basic recursion patterns.
// Usage: Run with `rustc factorial.rs && ./factorial`.
fn factorial(n: u64) -> u64 {
    match n {
        0 | 1 => 1,
        _ => n * factorial(n - 1),
    }
}

fn main() {
    let value = 6;
    println!("{}! = {}", value, factorial(value));
}
